### PR TITLE
Add ROI support to NvJpeg

### PR DESCRIPTION
**Pull request by szkarpinski**
## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:
This PR adds ROI support to NvJPEG decoder.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3016

### CODE REVIEW

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/decoders/nvjpeg/nvjpeg.cc [hunk 1]
```diff
@@ -165,14 +165,15 @@ DecodeResult NvJpegDecoderInstance::DecodeImplTask(int thread_idx,
...
+  }
   try {
-    if (roi) {
-      // TODO(msala) add support for ROI decoding
-      throw std::runtime_error("ROI support is not implemented yet");
-    }
-
     ParseJpegSample(*in, opts, ctx);
+    if (roi.use_roi()) {
+      ctx.shape = roi.shape();
```
Comment (staniewzki) [inline dali/imgcodec/decoders/nvjpeg/nvjpeg.cc:175]: I think that conditionally stripping channels off shape is error-prone.

Comment (szkarpinski) [reply dali/imgcodec/decoders/nvjpeg/nvjpeg.cc]: You're right, that's asking for trouble. Fixed, thanks!

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/decoders/nvjpeg/nvjpeg.cc [hunk 2]
```diff
@@ -165,14 +165,16 @@ DecodeResult NvJpegDecoderInstance::DecodeImplTask(int thread_idx,
   DecodingContext ctx = DecodingContext{ resources_[thread_idx] };
   CUDA_CALL(nvjpegDecodeParamsSetOutputFormat(ctx.resources.params, GetFormat(opts.format)));
   CUDA_CALL(nvjpegDecodeParamsSetAllowCMYK(ctx.resources.params, true));
-
+  if (roi.use_roi()) {
+    CUDA_CALL(nvjpegDecodeParamsSetROI(ctx.resources.params, roi.begin[1], roi.begin[0],
+                                       roi.shape()[1], roi.shape()[0]));
+  }
```
Comment (mzient) [inline dali/imgcodec/decoders/nvjpeg/nvjpeg.cc:171]: I think you should reset the params if not using ROI - unless ctx.resources.params is a disposable object.

```suggestion
  } else {
    CUDA_CALL(nvjpegDecodeParamsSetROI(ctx.resources.params, 0, 0, -1, -1));  }
```

Comment (szkarpinski) [reply dali/imgcodec/decoders/nvjpeg/nvjpeg.cc:173]: Wow, nice catch! Fixed

**Review by mzient — APPROVED** [review]

**Comment by szkarpinski** [comment]
> **CI MESSAGE**: [6013619]: BUILD FAILED

There were changes in the test framework recently and I didn't rebase. Fixed in 12c402c6b18b81a7a6c655e80040e0f59929f192 , restarting

**Comment by szkarpinski** [comment]
!build

**Pull request merged**